src/NuGet.Core/NuGet.Packaging/Signing/Authoring/Signer.cs

```diff
@@ -14,63 +14,80 @@ namespace NuGet.Packaging.Signing
     /// </summary>
     public sealed class Signer
     {
-        private readonly ISignedPackage _package;
         private readonly SigningSpecificationsV1 _specifications = SigningSpecifications.V1;
-        private readonly ISignatureProvider _signatureProvider;
+        private readonly SignerRequest _request;
 
         /// <summary>
         /// Creates a signer for a specific package.
         /// </summary>
         /// <param name="package">Package to sign or modify.</param>
-        public Signer(ISignedPackage package, ISignatureProvider signatureProvider)
+        public Signer(SignerRequest request)
         {
-            _package = package ?? throw new ArgumentNullException(nameof(package));
-            _signatureProvider = signatureProvider ?? throw new ArgumentNullException(nameof(signatureProvider));
+            _request = request ?? throw new ArgumentNullException(nameof(request));
         }
 
 #if IS_DESKTOP
         /// <summary>
         /// Add a signature to a package.
         /// </summary>
-        public async Task SignAsync(SignPackageRequest request, ILogger logger, CancellationToken token)
+        public async Task SignAsync(ILogger logger, CancellationToken token)
         {
-            if (request == null)
-            {
-                throw new ArgumentNullException(nameof(request));
-            }
-
             if (logger == null)
             {
                 throw new ArgumentNullException(nameof(logger));
             }
 
             token.ThrowIfCancellationRequested();
 
-            if (await _package.IsZip64Async(token))
+            SigningUtility.Verify(_request.SignRequest, logger);
+
+            var inputPackagePath = _request.PackagePath;
+            var tempPackagePath = Path.GetTempFileName();
+
+            using (var packageReadStream = File.OpenRead(inputPackagePath))
+            using (var packageWriteStream = File.Open(tempPackagePath, FileMode.OpenOrCreate))
+            using (var package = new SignedPackageArchive(packageReadStream, packageWriteStream))
             {
-                throw new SignatureException(NuGetLogCode.NU3006, Strings.ErrorZip64NotSupported);
+                var primarySignature = await package.GetPrimarySignatureAsync(token);
+
+                if (_request.Overwrite && primarySignature != null)
+                {
+                    await package.RemoveSignatureAsync(token);
+                    inputPackagePath = tempPackagePath;
+                }
             }
 
-            SigningUtility.Verify(request, logger);
+            using (var packageReadStream = File.OpenRead(inputPackagePath))
```

Comment:
replace all of this with relevant safer api implementations from FileUtility class